### PR TITLE
Sam speakerloss

### PR DESCRIPTION
**Pull request by sjkoelle**

### CODE REVIEW

File: uberduck_ml_dev/text/symbols.py [hunk 1]
```diff
@@ -13,44 +13,62 @@
...
 
 # Prepend "@" to ARPAbet symbols to ensure uniqueness (some are the same as
 # uppercase letters):
 _arpabet = ["@" + s for s in cmudict.valid_symbols]
 
 # Export all symbols:
 symbols = (
-    list(_punctuation + _math + _special + _accented + _numbers + _letters) + _arpabet
+    list(_pad + _punctuation + _math + _special + _accented + _numbers + _letters)
+    + _arpabet
```
Comment (zachwe) [inline uberduck_ml_dev/text/symbols.py:32]: Just FYI, this will break pretrained models due to changing the symbol sets.

File: uberduck_ml_dev/trainer/base.py [hunk 2]
```diff
@@ -140,16 +141,21 @@ def forward(self, model_output: List, target: List):
         mel_target, gate_target = target[0], target[1]
         mel_target.requires_grad = False
         gate_target.requires_grad = False
-        gate_target = gate_target.view(-1, 1)
+        # gate_target = gate_target.view(-1, 1)
```
Comment (zachwe) [inline uberduck_ml_dev/trainer/base.py:144]: delete

File: uberduck_ml_dev/trainer/base.py [hunk 3]
```diff
@@ -193,6 +201,27 @@ def log_training(
             scalar=step_duration_seconds,
         )
 
+        batch_levels = X[5]
+        batch_levels_unique = torch.unique(batch_levels)
+        for l in batch_levels_unique:
+            mlb = mel_loss_batch[torch.where(batch_levels == l)[0]].mean()
+            self.log(
+                "MelLoss/train/" + str(l),
```
Comment (zachwe) [inline uberduck_ml_dev/trainer/base.py:209]: Here and in the two other cases below, could we do something like `f"MelLoss/train/speaker{l.item()}"` so that the graphs are labeled with names like `speaker0` instead of `tensor(0, device="cuda:0")`

File: uberduck_ml_dev/trainer/base.py [hunk 4]
```diff
@@ -557,8 +596,11 @@ def train(self):
                 else:
                     reduced_mel_loss = mel_loss.item()
                     reduced_gate_loss = gate_loss.item()
-                reduced_loss = reduced_mel_loss + reduced_gate_loss
+                    reduced_gate_loss_batch = gate_loss_batch  # .item()
+                    reduced_mel_loss_batch = mel_loss_batch  # .item()
```
Comment (zachwe) [inline uberduck_ml_dev/trainer/base.py:600]: Should we call `.detach()` on these? I think otherwise we'll compute extra gradients.

**Comment by zachwe** [comment]
Looks like tests are failing due to `phonemizer` (and a periodic flake in downloading NLTK stuff, ugh), feel free to add it or I will on the VITS branch.

File: uberduck_ml_dev/trainer/base.py (hunk 3 above)
Comment (sjkoelle) [reply uberduck_ml_dev/trainer/base.py:209]: good point

File: uberduck_ml_dev/trainer/base.py (hunk 4 above)
Comment (sjkoelle) [reply uberduck_ml_dev/trainer/base.py:600]: I don't think we will have this issue?  We call loss.backward() where loss = gate_loss + mel_loss, so I'm not sure adding outputs that we don't differentiate through will add gradients.  Why not detach tho

**Review by zachwe — APPROVED** [review]
I'll figure out the tests when I merge in the remaining VITS stuff
